engine/grid.py

```diff
@@ -301,79 +320,379 @@ def find_border_nodes(self):
+            # let's traverse this last row
+            if self.direction == self.Direction.RIGHT:
+                node_info_curr = max(candidate_nodes_curr, key=lambda node_info: node_info[1])
+            elif self.direction == self.Direction.LEFT:
+                node_info_curr = min(candidate_nodes_curr, key=lambda node_info: node_info[1])
+            elif self.direction == self.Direction.UP:
+                node_info_curr = max(candidate_nodes_curr, key=lambda node_info: node_info[2])
+            elif self.direction == self.Direction.DOWN:
+                node_info_curr = min(candidate_nodes_curr, key=lambda node_info: node_info[2])
+            
+            if self.direction == self.Direction.RIGHT or self.direction == self.Direction.LEFT:
+                return node_info_curr[1]
+            else: 
+                return node_info_curr[2]
         else:
-            node_info = min(candidate_nodes,key=lambda node_info: node_info[1])
-            return (node_info[1],node_info[2])
-
-
+            if self.direction == self.Direction.RIGHT:
+                node_info_next = max(candidate_nodes_next, key=lambda node_info: node_info[1])
+                node_info_curr = max(candidate_nodes_curr, key=lambda node_info: node_info[1])
+                return min(node_info_next[1], node_info_curr[1])
+            elif self.direction == self.Direction.LEFT:
+                node_info_next = min(candidate_nodes_next, key=lambda node_info: node_info[1])
+                node_info_curr = min(candidate_nodes_curr, key=lambda node_info: node_info[1])
+                return max(node_info_next[1], node_info_curr[1])
+            elif self.direction == self.Direction.UP:
+                node_info_next = max(candidate_nodes_next, key=lambda node_info: node_info[2])
+                node_info_curr = max(candidate_nodes_curr, key=lambda node_info: node_info[2])
+                return min(node_info_next[2], node_info_curr[2])
+            elif self.direction == self.Direction.DOWN:
+                node_info_next = min(candidate_nodes_next, key=lambda node_info: node_info[2])
+                node_info_curr = min(candidate_nodes_curr, key=lambda node_info: node_info[2])
+                return max(node_info_next[2], node_info_curr[2])
+
+    def plot_circle(self, start_pos, end_pos, center, orientation, theta_step=math.pi / 12):
+        """
+        Returns a circle of nodes starting from the [start_pos], going at orientation [orientation], and ending at the
+        [end_pos] with center [center].
+
+        Arguments:
+            start_pos: float tuple representing the starting point of the circle being plotted
+            end_pos: float tuple representing the ending point of the circle being plotted
+            center: float tuple representing the center point of the circle being plotted
+            orientation: orientation that the nodes are being plotted from, starting with the start_pos and ending
+                at the end_pos
```

Comment:
lmao, what happened here. why is it on another line

Reply:
🤷